webapp-django/crashstats/api/tests/test_views.py

```diff
@@ -84,65 +82,87 @@ def test_base_classes_raise_not_found(self):
         assert 'max-age={}'.format(cache_seconds) in response['Cache-Control']
 
+    def test_metrics_gathering(self):
+        url = reverse('api:model_wrapper', args=('NoOp',))
+        with MetricsMock() as metrics_mock:
+            response = self.client.get(url, {'product': 'good'})
+        assert response.status_code == 200
+        assert metrics_mock.has_record(
+            fun_name='incr',
+            stat='webapp.api.pageview',
+            value=1,
+            tags=['endpoint:apiNoOp']
+        )
+
+    def test_param_exceptions(self):
+        # missing required parameter
+        url = reverse('api:model_wrapper', args=('NoOp',))
+        response = self.client.get(url)
+        assert response.status_code == 400
+        assert 'This field is required.' in response.content
+
+        response = self.client.get(url, {'product': 'bad'})
+        assert response.status_code == 400
+        assert 'Bad value for parameter(s) \'Bad product\'' in response.content
+
+    def test_hit_or_not_hit_ratelimit(self):
+        url = reverse('api:model_wrapper', args=('NoOp',))
+
+        response = self.client.get(url, {'product': 'good'})
+        assert response.status_code == 200
+        with self.settings(
+            API_RATE_LIMIT='3/m',
+            API_RATE_LIMIT_AUTHENTICATED='6/m'
+        ):
+            current_limit = 3  # see above mentioned settings override
+            # Double to avoid
+            # https://bugzilla.mozilla.org/show_bug.cgi?id=1148470
+            for i in range(current_limit * 2):
+                response = self.client.get(url, {'product': 'good'}, HTTP_X_REAL_IP='12.12.12.12')
+            assert response.status_code == 429
+
+            # But it'll work if you use a different X-Real-IP
+            # because the rate limit is based on your IP address
+            response = self.client.get(url, {'product': 'good'}, HTTP_X_REAL_IP='11.11.11.11')
+            assert response.status_code == 200
+
+            user = User.objects.create(username='test')
+            token = Token.objects.create(
+                user=user,
+                notes="Just for avoiding rate limit"
+            )
+
+            response = self.client.get(url, {'product': 'good'}, HTTP_AUTH_TOKEN=token.key)
+            assert response.status_code == 200
+
+            for i in range(current_limit):
+                response = self.client.get(url, {'product': 'good'})
+            assert response.status_code == 200
+
+            # But even being signed in has a limit.
+            authenticated_limit = 6  # see above mentioned settings override
+            assert authenticated_limit > current_limit
+            for i in range(authenticated_limit * 2):
+                response = self.client.get(url, {'product': 'good'})
+            # Even if you're authenticated - sure the limit is higher -
+            # eventually you'll run into the limit there too.
+            assert response.status_code == 429
+
```

Comment:
I moved all the api infrastructure tests to the same part of the class and the middleware tests after it. So this code is the same as what we had below, but it was in a different part of the file. The only change is that it now uses the `NoOpMiddleware`.